contracts/contracts/FeeSub.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/token/ERC20/IERC20.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/token/ERC20/utils/SafeERC20.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/access/Ownable.sol";
+
+import "./RequestManager.sol";
+
+contract FeeSub is Ownable {
+    using SafeERC20 for IERC20;
+
+    RequestManager public requestManager;
+    mapping(address token => uint256) public minimumAmounts;
+    mapping(bytes32 requestId => address sender) public senders;
+
+    constructor(address _requestManager) {
+        requestManager = RequestManager(_requestManager);
+    }
+
+    function createRequest(
+        uint256 targetChainId,
+        address sourceTokenAddress,
+        address targetTokenAddress,
+        address targetAddress,
+        uint256 amount,
+        uint256 validityPeriod
+    ) external returns (bytes32) {
+        require(
+            minimumAmounts[sourceTokenAddress] > 0,
+            "Token not to be subsidized"
+        );
+        require(
+            amount >= minimumAmounts[sourceTokenAddress],
+            "Transfer amount too small to be subsidized"
+        );
+
+        require(
+            IERC20(sourceTokenAddress).allowance(msg.sender, address(this)) >=
+                amount,
+            "Insufficient allowance"
+        );
+
+        IERC20(sourceTokenAddress).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amount
+        );
+
+        bytes32 requestId = requestManager.createRequest(
+            targetChainId,
+            sourceTokenAddress,
+            targetTokenAddress,
+            targetAddress,
+            amount,
+            validityPeriod
+        );
+        senders[requestId] = msg.sender;
+
+        return requestId;
+    }
+
+    function withdrawExpiredRequest(bytes32 requestId) external {
+        (
+            address sender,
+            address sourceTokenAddress,
+            ,
+            uint256 amount,
+            ,
+            ,
+            ,
+            ,
+            uint96 withdrawClaimId,
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already withdrawn");
```

Comment:
I would rephrase it to "already refunded to the sender".
The rationale behind this is that we should have a distinction between withdrawing from the request manager and withdrawing from this contract.

Below there is `Already withdrawn by agent` -> I would this rename to `request is withdrawn by another address` (wouldnt mention the notion agent here as agent refers more to the software component)